airbrakes/data_handling/apogee_predictor.py

```diff
@@ -188,12 +213,38 @@ def _prediction_loop(self) -> None:
                 self._current_altitude = data_packet.current_altitude
                 self._current_velocity = data_packet.vertical_velocity
 
-            # TODO: play around with this value
-            # Run apogee prediction every 100 data points:
             if len(self._accelerations) - last_run_length >= APOGEE_PREDICTION_FREQUENCY:
+                # We only want to keep curve fitting if the curve fit hasn't converged yet
+                if not self._has_apogee_converged():
+                    curve_coefficients = self._curve_fit()
+                    lookup_table = self._generate_prediction(curve_coefficients)
                 self._cumulative_time_differences = np.cumsum(self._time_differences)
-                params = self._curve_fit()
-                self._apogee_prediction_value.value = self._get_apogee(params)
+                # Predicts the apogee using the lookup table and linear interpolation
+                predicted_apogee = np.interp(
+                    self._current_velocity, lookup_table[0], lookup_table[1]
+                )
+                self._apogee_prediction_value.value = predicted_apogee
+                self._predicted_apogees = np.append(self._predicted_apogees, predicted_apogee)
                 last_run_length = len(self._accelerations)
             # notifies tests that prediction is complete
             self._prediction_complete.set()
+
+    def _has_apogee_converged(self) -> bool:
```

Comment:
ok I just had a deeper look on this. We should be checking apogee convergence not like this, but by using the covariance matrix returned by the curve_fit function. 

Currently with this approach, the code says that apogee has converged below 1800m for interest launch, which is wrong. A simple check like this is better imo (less code, faster):

```python
popt, pcov = scipy.optimize.curve_fit(func, xdata, ydata)
uncertainties = np.sqrt(np.diag(pcov))

if np.all(uncertainties < threshold):
    print("Fit has converged")
```

I found a good threshold to be: 
```
threshold = np.array([[0.0259, 0.00065]])
```
which works for both interest and purple launch.

Reply:
another alternative would be to use the standard deviation, however this is not without limitations also, since it is susceptible to local maximas of our predicted apogee graph, so it can converge early. If you use these settings:

Threshold: <= 1.7, `NUMBER_OF_PREDICITIONS = 10`, then:

converged at 1868m for int launch, at alt = 614m
converged at 795m for purple launch, at alt = 496m

Reply:
okay we can add this but you'll have to explain to me how tf it works